internal/expr/prepare.go

```diff
@@ -145,54 +123,127 @@ func prepareOutput(ti typeNameToInfo, p *outputPart) (outCols []fullName, typeMe
 					return nil, nil, err
 				}
 				typeMembers = append(typeMembers, tm)
-				outCols = append(outCols, fullName{pref, t.name})
+				genCols = append(genCols, fullName{pref, t.name})
 			}
 		}
-		return outCols, typeMembers, nil
+		return genCols, typeMembers, nil
 	} else if numColumns > 1 && starColumns > 0 {
 		return nil, nil, fmt.Errorf("invalid asterisk in columns")
 	}
 
-	// Case 2: Explicit columns, single asterisk type e.g. "(col1, t.col2) AS &P.*".
+	// Case 2: Explicit columns, single asterisk type.
+	// For example the expressions:
+	//  "(col1, col2) VALUES ($P.*)"
+	//  "(col1, t.col2) AS (&P.*)"
 	if starTypes == 1 && numTypes == 1 {
-		if info, err = fetchInfo(p.targetTypes[0].prefix); err != nil {
+		info, err := ti.lookupInfo(types[0].prefix)
+		if err != nil {
 			return nil, nil, err
 		}
-		for _, c := range p.sourceColumns {
+		for _, c := range columns {
 			tm, err := info.typeMember(c.name)
 			if err != nil {
 				return nil, nil, err
 			}
 			typeMembers = append(typeMembers, tm)
-			outCols = append(outCols, c)
+			genCols = append(genCols, c)
 		}
-		return outCols, typeMembers, nil
+		return genCols, typeMembers, nil
 	} else if starTypes > 0 && numTypes > 1 {
 		return nil, nil, fmt.Errorf("invalid asterisk in types")
 	}
 
-	// Case 3: Explicit columns and types e.g. "(col1, col2) AS (&P.name, &P.id)".
+	// Case 3: Explicit columns and types.
+	// For example the expressions:
+	//  "(col1, col2) VALUES ($P.name, $A.id)"
+	//  "(col1, col2) AS (&P.name, &P.id)"
 	if numColumns == numTypes {
-		for i, c := range p.sourceColumns {
-			t := p.targetTypes[i]
-			if info, err = fetchInfo(t.prefix); err != nil {
+		for i, c := range columns {
+			t := types[i]
+			info, err := ti.lookupInfo(t.prefix)
+			if err != nil {
 				return nil, nil, err
 			}
 			tm, err := info.typeMember(t.name)
 			if err != nil {
 				return nil, nil, err
 			}
 			typeMembers = append(typeMembers, tm)
-			outCols = append(outCols, c)
+			genCols = append(genCols, c)
 		}
 	} else {
 		return nil, nil, fmt.Errorf("mismatched number of columns and target types")
 	}
+	return genCols, typeMembers, nil
+}
 
-	return outCols, typeMembers, nil
+// prepareOutput checks that the output expressions correspond to known types.
+// It then checks they are formatted correctly and finally generates the columns for the query.
+func prepareOutput(ti typeNameToInfo, p *outputPart) ([]fullName, []typeMember, error) {
+	outCols, typeMembers, err := prepareColumnsAndTypes(ti, p.sourceColumns, p.targetTypes)
+	if err != nil {
+		err = fmt.Errorf("output expression: %s: %s", err, p.raw)
+	}
+	return outCols, typeMembers, err
 }
 
-type typeNameToInfo map[string]typeInfo
+// isInsert returns true if the input expression inside an INSERT statement.
+func isInsert(p *inputPart) bool {
+	return len(p.targetColumns) > 0
+}
```

Comment:
Have you thought about putting this as a method of `inputPart` ? To me, it is more clear where the information about the struct is all together in its methods.

Reply:
Yea, I suppose it would go best there.